File: src/main/java/platform/ecommerce/domain/product/Product.java
package platform.ecommerce.domain.product;

import jakarta.persistence.*;
import lombok.*;
import platform.ecommerce.domain.common.BaseEntity;
import platform.ecommerce.domain.common.SoftDeletable;
import platform.ecommerce.exception.ErrorCode;
import platform.ecommerce.exception.InvalidStateException;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

/**
 * Product aggregate root.
 * Manages product options, images, and lifecycle.
 */
@Entity
@Table(name = "product")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Product extends BaseEntity implements SoftDeletable {

    private static final int MAX_OPTIONS = 50;
    private static final int MAX_IMAGES = 20;

    @Column(nullable = false, length = 200)
    private String name;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(name = "base_price", nullable = false, precision = 12, scale = 2)
    private BigDecimal basePrice;

    @Column(name = "seller_id", nullable = false)
    private Long sellerId;

    @Column(name = "category_id")
    private Long categoryId;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private ProductStatus status;

    @Column(name = "deleted_at")
    private LocalDateTime deletedAt;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    @OrderBy("displayOrder ASC")
    private List<ProductOption> options = new ArrayList<>();

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    @OrderBy("displayOrder ASC")
    private List<ProductImage> images = new ArrayList<>();

    @Builder
    public Product(String name, String description, BigDecimal basePrice, Long sellerId, Long categoryId) {
        validateRequired(name, "name");
        validatePrice(basePrice);

        this.name = name;
        this.description = description;
        this.basePrice = basePrice;
        this.sellerId = sellerId;
        this.categoryId = categoryId;
        this.status = ProductStatus.DRAFT;
    }

    // ========== Status Transitions ==========

    /**
     * Publish product for sale.
     */
    public void publish() {
        if (this.status != ProductStatus.DRAFT) {
            throw new InvalidStateException(ErrorCode.PRODUCT_NOT_AVAILABLE, "Only draft products can be published");
        }
        if (this.options.isEmpty()) {
            throw new InvalidStateException(ErrorCode.PRODUCT_NOT_AVAILABLE, "Product must have at least one option");
        }
        this.status = ProductStatus.ACTIVE;
    }

    /**
     * Discontinue product.
     */
    public void discontinue() {
        if (this.status == ProductStatus.DISCONTINUED) {
            throw new InvalidStateException(ErrorCode.PRODUCT_NOT_AVAILABLE, "Product is already discontinued");
        }
        this.status = ProductStatus.DISCONTINUED;
    }

    /**
     * Check and update status based on stock.
     */
    public void updateStatusByStock() {
        if (this.status == ProductStatus.ACTIVE && getTotalStock() == 0) {
            this.status = ProductStatus.SOLD_OUT;
        } else if (this.status == ProductStatus.SOLD_OUT && getTotalStock() > 0) {
            this.status = ProductStatus.ACTIVE;
        }
    }

    // ========== Product Update ==========

    /**
     * Update product details.
     */
    public void update(String name, String description, BigDecimal basePrice, Long categoryId) {
        if (name != null && !name.isBlank()) {
            this.name = name;
        }
        this.description = description;
        if (basePrice != null) {
            validatePrice(basePrice);
            this.basePrice = basePrice;
        }
        this.categoryId = categoryId;
    }

    // ========== Option Management ==========

    /**
     * Add option to product.
     */
    public ProductOption addOption(OptionType type, String value, BigDecimal additionalPrice, int stock) {
        if (this.options.size() >= MAX_OPTIONS) {
            throw new InvalidStateException(ErrorCode.INVALID_INPUT, "Maximum options limit reached");
        }

        int displayOrder = this.options.size();
        ProductOption option = ProductOption.builder()
                .product(this)
                .optionType(type)
                .optionValue(value)
                .additionalPrice(additionalPrice)
                .stock(stock)
                .displayOrder(displayOrder)
                .build();

        this.options.add(option);
        return option;
    }

    /**
     * Remove option from product.
     */
    public void removeOption(Long optionId) {
        ProductOption option = findOptionById(optionId);
        this.options.remove(option);
        updateStatusByStock();
    }

    /**
     * Find option by ID.
     */
    public ProductOption findOptionById(Long optionId) {
        return this.options.stream()
                .filter(o -> o.getId().equals(optionId))
                .findFirst()
                .orElseThrow(() -> new InvalidStateException(ErrorCode.PRODUCT_OPTION_NOT_FOUND));
    }

    /**
     * Get total stock across all options.
     */
    public int getTotalStock() {
        return this.options.stream()
                .mapToInt(ProductOption::getStock)
                .sum();
    }

    /**
     * Decrease stock for specific option.
     */
    public void decreaseStock(Long optionId, int quantity) {
        ProductOption option = findOptionById(optionId);
        option.decreaseStock(quantity);
        updateStatusByStock();
    }

    /**
     * Increase stock for specific option.
     */
    public void increaseStock(Long optionId, int quantity) {
        ProductOption option = findOptionById(optionId);
        option.increaseStock(quantity);
        updateStatusByStock();
    }

    // ========== Image Management ==========

    /**
     * Add image to product.
     */
    public ProductImage addImage(String imageUrl, String altText) {
        if (this.images.size() >= MAX_IMAGES) {
            throw new InvalidStateException(ErrorCode.INVALID_INPUT, "Maximum images limit reached");
        }

        int displayOrder = this.images.size();
        ProductImage image = ProductImage.builder()
                .product(this)
                .imageUrl(imageUrl)
                .altText(altText)
                .displayOrder(displayOrder)
                .build();

        this.images.add(image);
        return image;
    }

    /**
     * Remove image from product.
     */
    public void removeImage(Long imageId) {
        ProductImage image = findImageById(imageId);
        this.images.remove(image);
        reorderImages();
    }

    /**
     * Find image by ID.
     */
    public ProductImage findImageById(Long imageId) {
        return this.images.stream()
                .filter(i -> i.getId().equals(imageId))
                .findFirst()
                .orElseThrow(() -> new InvalidStateException(ErrorCode.PRODUCT_IMAGE_NOT_FOUND));
    }

    /**
     * Get main image (first in order).
     */
    public ProductImage getMainImage() {
        return this.images.stream()
                .min(Comparator.comparingInt(ProductImage::getDisplayOrder))
                .orElse(null);
    }

    /**
     * Reorder images after removal.
     */
    private void reorderImages() {
        List<ProductImage> sorted = new ArrayList<>(this.images);
        sorted.sort(Comparator.comparingInt(ProductImage::getDisplayOrder));
        for (int i = 0; i < sorted.size(); i++) {
            sorted.get(i).updateOrder(i);
        }
    }

    // ========== Soft Delete ==========

    @Override
    public void delete() {
        this.deletedAt = LocalDateTime.now();
        this.status = ProductStatus.DISCONTINUED;
    }

    @Override
    public void restore() {
        if (this.deletedAt == null) {
            throw new InvalidStateException(ErrorCode.INVALID_INPUT, "Product is not deleted");
        }
        this.deletedAt = null;
        this.status = ProductStatus.DRAFT;
    }

    // ========== Validation ==========

    private void validateRequired(String value, String fieldName) {
        if (value == null || value.isBlank()) {
            throw new IllegalArgumentException(fieldName + " is required");
        }
    }

    private void validatePrice(BigDecimal price) {
        if (price == null || price.compareTo(BigDecimal.ZERO) < 0) {
            throw new IllegalArgumentException("Price cannot be negative");
        }
    }
}

File: src/main/kotlin/platform/ecommerce/domain/Product.kt
package platform.ecommerce.domain

import jakarta.persistence.*
import jakarta.persistence.GenerationType.IDENTITY
import org.hibernate.annotations.SQLRestriction
import platform.ecommerce.domain.vo.Money
import platform.ecommerce.enums.ProductStatus
import platform.ecommerce.utils.SkuGenerator.generateOptionSku

@Entity
@SQLRestriction("deleted_at IS NULL")
class Product(
    @Column(nullable = false, unique = true, length = 100)
    var sku: String,

    @Column(nullable = false)
    var name: String,

    @Column(columnDefinition = "TEXT")
    var description: String,

    @Embedded
    private var price: Money,

    @Column(length = 500, name = "image_url")
    var imageUrl: String,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, name = "product_status")
    private var productStatus: ProductStatus = ProductStatus.AVAILABLE
): SoftDeletableEntity() {

    @Id @GeneratedValue(strategy = IDENTITY)
    var id: Long? = null
        protected set

    @OneToMany(mappedBy = "product", cascade = [CascadeType.ALL], orphanRemoval = true)
    private val productOptions: MutableList<ProductOption> = mutableListOf()

    val options: List<ProductOption>
        get() = productOptions.toList()

    fun getPrice(): Money = price

    fun getProductStatus(): ProductStatus = productStatus

    fun changePrice(newPrice: Money) {
        require(productStatus != ProductStatus.DISCONTINUED) {
            "Cannot change price of discontinued product"
        }
        this.price = newPrice
    }

    fun updateInfo(name: String, description: String, imageUrl: String) {
        require(name.isNotBlank()) { "Name cannot be blank" }
        this.name = name
        this.description = description
        this.imageUrl = imageUrl
    }

    fun discontinue() {
        this.productStatus = ProductStatus.DISCONTINUED
    }

    fun isAvailable(): Boolean {
        return productStatus == ProductStatus.AVAILABLE && hasAvailableStock()
    }

    fun addOption(optionName: String, stockQuantity: Int) {
        val optionSku = generateOptionSku(this.sku, optionName)
        val option = ProductOption.create(
            product = this,
            sku = optionSku,
            optionName = optionName,
            stockQuantity = stockQuantity
        )
        productOptions.add(option)
        checkAndUpdateStatus()
    }

    fun removeOption(optionId: Long) {
        val option = productOptions.find { it.id == optionId }
            ?: throw IllegalArgumentException("Product option with ID $optionId not found")
        productOptions.remove(option)
        checkAndUpdateStatus()
    }

    internal fun checkAndUpdateStatus() {
        if (productStatus == ProductStatus.DISCONTINUED) {
            return
        }

        productStatus = if (hasAvailableStock()) {
            ProductStatus.AVAILABLE
        } else {
            ProductStatus.OUT_OF_STOCK
        }
    }

    private fun hasAvailableStock(): Boolean {
        return productOptions.any { it.isOrderable() }
    }

    companion object {
        fun create(
            sku: String,
            name: String,
            description: String,
            price: Double,
            imageUrl: String
        ): Product {
            require(sku.isNotBlank()) { "SKU cannot be blank" }
            require(name.isNotBlank()) { "Name cannot be blank" }
            require(price >= 0) { "Price must be non-negative" }

            return Product(
                sku = sku,
                name = name,
                description = description,
                price = Money.of(price),
                imageUrl = imageUrl
            )
        }
    }
}